File: Source/typescript/features/FeaturesParser.ts
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { BooleanFeatureToggle } from './BooleanFeatureToggle';
import { Feature } from './Feature';
import { FeatureDefinitions } from './FeatureDefinitions';
import { Features } from './Features';
import { IFeaturesParser } from './IFeaturesParser';

/**
 * Represents an implementation of {@link IFeaturesParser}.
 */
export class FeaturesParser implements IFeaturesParser {

    /** @inheritdoc */
    parse(json: string): Features {
        const featuresDefinitions = JSON.parse(json) as FeatureDefinitions;
        const features = new Map<string, Feature>();
        for (const featureName in featuresDefinitions) {
            const featureDefinition = featuresDefinitions[featureName];

            features.set(featureName, new Feature(
                featureName,
                featureDefinition.description,
                featureDefinition.toggles.map(_ => new BooleanFeatureToggle(_.isOn))));
        }

        return new Features(features);
    }
}

File: Source/DotNET/Backend/Features/FeaturesParser.cs
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;

namespace Dolittle.Vanir.Backend.Features
{
    /// <summary>
    /// Represents an implementation of <see cref="IFeaturesParser"/>
    /// </summary>
    public class FeaturesParser : IFeaturesParser
    {
        /// <inheritdoc/>
        public Features Parse(string json)
        {
            var featureDefinitions = JsonConvert.DeserializeObject<Dictionary<string, FeatureDefinition>>(json);
            var features = featureDefinitions.ToDictionary(
                _ => _.Key,
                _ => new Feature
                {
                    Name = _.Key,
                    Description = _.Value.Description,
                    Toggles = _.Value.Toggles.Select(t => new BooleanFeatureToggle { IsOn = t.IsOn })
                });
            return new Features(features);
        }
    }
}
